langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiStreamingChatModel.java

```diff
@@ -0,0 +1,228 @@
+    private final String apiKey;
+    private final String modelName;
+    private final Double temperature;
+    private final Integer topK;
+    private final Double topP;
+    private final Integer maxOutputTokens;
+    private final List<String> stopSequences;
+    private final Integer candidateCount;
+    private final ResponseFormat responseFormat;
+    private final GeminiFunctionCallingConfig toolConfig;
+    private final boolean allowCodeExecution;
+    private final boolean includeCodeExecutionOutput;
+    private final List<GeminiSafetySetting> safetySettings;
+    private final List<ChatModelListener> listeners;
+    private final Integer maxRetries;
+
+    @Builder
+    public GoogleAiGeminiStreamingChatModel(String apiKey, String modelName,
+                                            Double temperature, Integer topK, Double topP,
+                                            Integer maxOutputTokens, Duration timeout,
+                                            ResponseFormat responseFormat,
+                                            List<String> stopSequences, GeminiFunctionCallingConfig toolConfig,
+                                            Boolean allowCodeExecution, Boolean includeCodeExecutionOutput,
+                                            Boolean logRequestsAndResponses,
+                                            List<GeminiSafetySetting> safetySettings,
+                                            List<ChatModelListener> listeners,
+                                            Integer maxRetries) {
+        this.apiKey = ensureNotBlank(apiKey, "apiKey");
+        this.modelName = ensureNotBlank(modelName, "modelName");
+        this.temperature = temperature;
+        this.topK = topK;
+        this.topP = topP;
+        this.maxOutputTokens = maxOutputTokens;
+        this.candidateCount = 1; // Multiple candidates aren't supported by langchain4j
+        this.stopSequences = getOrDefault(stopSequences, emptyList());
+        this.toolConfig = toolConfig;
+        this.allowCodeExecution = getOrDefault(allowCodeExecution, false);
+        this.includeCodeExecutionOutput = getOrDefault(includeCodeExecutionOutput, false);
+        this.safetySettings = copyIfNotNull(safetySettings);
+        this.responseFormat = responseFormat;
+        this.listeners = listeners == null ? emptyList() : new ArrayList<>(listeners);
+        this.maxRetries = getOrDefault(maxRetries, 3);
+        this.geminiService = new GeminiService(
+                getOrDefault(logRequestsAndResponses, false) ? log : null,
+                getOrDefault(timeout, ofSeconds(60))
+        );
+    }
+
+    @Override
+    public void generate(List<ChatMessage> messages, StreamingResponseHandler<AiMessage> handler) {
+        generate(messages, Collections.emptyList(), handler);
+    }
+
+    @Override
+    public void generate(List<ChatMessage> messages, ToolSpecification toolSpecification, StreamingResponseHandler<AiMessage> handler) {
```

Comment:
Should this then also be the same for the `GoogleGeminiAiChatModel`?

Reply:
@Bjarne-Kinkel yes

Reply:
Okay, but there is a problem. The method in the `GoogleGeminiAiChatModel` is already incorrectly implemented. However, Gemini doesn’t offer the option to enforce a tool call. Ideally, the method should throw an exception. Since this function is already implemented incorrectly, changing it would result in a breaking change. Should I still modify the method to throw an exception?

Reply:
Hmm, it seems that tool call can be enforced by using `mode=ANY` and `allowed_function_names=[{my_tool}]`:
https://ai.google.dev/gemini-api/docs/function-calling?authuser=1#function_calling_mode

Reply:
Yes, if we really want to follow the principle that the generate method with a single call means forcing a tool call, then we should use the ANY mode to do that.

But I still find a bit unfortunate that this generate method variant exist, as it was modeled after a specific behavior of OpenAI GPTs. It's not generic and forces all other models to somehow behave like this even if they are not supporting this approach. The name of the method should have been different and should have been explicit about its behavior.

Note that the Vertex AI Gemini behaves as if a list of one tool was passed. So the tool choice is not enforced.
I wonder if it's okay to have the two models behave differently now.

Reply:
@glaforge I agree that the name is bad, although I disagree about the existence of the method itself. Structured data extraction is a big use case and function calling was used for it for a long time. So there is definitely a need to force the model to call a specific tool. Most LLM providers (OpenAI, Gemini, Bedrock, Anthropic, Mistral, etc) support this behaviour, so I see no problem with this.

> Note that the Vertex AI Gemini behaves as if a list of one tool was passed. So the tool choice is not enforced.
> I wonder if it's okay to have the two models behave differently now.

Oh, I was not aware of this, we will need to fix this. But I guess we can just wait till we migrate to the new `chat(ChatRequest)` API, `ChatRequest` weill eventually have something like `ToolMode`

Reply:
@Bjarne-Kinkel I guess it makes sense to keep it separate and just remove this method for now (so that default impl will take care of throwing exception). Once the new API will be in place, we can implement this behaviour in a separate PR

Reply:
Is that for the normal ChatModel implementation as well or just for my streaming version?

Reply:
let's keep `GoogleAiGeminiChatModel` as it is now and only remove this method from `GoogleAiGeminiStreamingChatModel`

Reply:
Okay, perfect. Done.